ocaml/xapi/xapi_host_driver.ml

```diff
@@ -0,0 +1,297 @@
+        )
+      )
+      (split a) (split b)
+end
+
+let get_driver_versions driver_name selected_dir all_versions_dir =
+  let selected_version =
+    try
+      let driver_realpath =
+        Unix.readlink (selected_dir // (driver_name ^ ".ko"))
+      in
+      Filename.(basename (dirname driver_realpath))
+    with _ ->
+      D.debug "Driver '%s' has no selected version" driver_name ;
+      ""
+  in
+  let active_version =
+    try
+      Result.value ~default:""
+        (Xapi_host_driver_helpers.get_version
+           ("/sys/module" // driver_name // "notes/note.xenserver")
+        )
+    with _ ->
+      D.debug "Driver '%s' has no active version" driver_name ;
+      ""
+  in
+  let driver_versions =
+    let arr = Sys.readdir (all_versions_dir // driver_name) in
+    Array.sort Version.compare arr ;
+    Array.to_list arr
+  in
+  (selected_version, active_version, driver_versions)
+
+(* Scans the multi-version drivers filesystem hierarchy, updating the drivers
+   already in the DB if necessary, removing the ones that have been deleted
+   from the filesystem, and adding new ones *)
+let discover ~__context ~host =
+  try
+    let driver_paths = get_drivers_base_path ~__context ~host in
+    let db_drivers_set, db_drivers_map =
+      Db.Host_driver.get_all_records ~__context
+      |> List.fold_left
+           (fun (set, map) (ref, driver) ->
+             if driver.API.host_driver_host = host then
+               ( DriverSet.add driver.API.host_driver_name_label set
+               , DriverMap.add driver.API.host_driver_name_label ref map
+               )
+             else
+               (set, map)
+           )
+           (DriverSet.empty, DriverMap.empty)
+    in
+    let filesystem_drivers =
+      Array.fold_left
+        (fun set name -> DriverSet.add name set)
+        DriverSet.empty
+        (Sys.readdir driver_paths.all_versions_dir)
+    in
+    (* Modify drivers that were in the DB before and are still on the
+       filesystem in-place to keep references from outside intact *)
+    let updated_drivers =
+      DriverSet.inter db_drivers_set filesystem_drivers
+      |> DriverSet.elements
+      |> List.filter_map (fun driver_name ->
+             let self = DriverMap.find driver_name db_drivers_map in
+             let selected_version, active_version, driver_versions =
+               get_driver_versions driver_name driver_paths.selected_dir
+                 driver_paths.all_versions_dir
+             in
+             let selected_changed =
+               Db.Host_driver.get_selected_version ~__context ~self
+               <> selected_version
+             in
+             let active_changed =
+               Db.Host_driver.get_active_version ~__context ~self
+               <> active_version
+             in
+             let versions_changed =
+               Db.Host_driver.get_versions ~__context ~self <> driver_versions
+             in
+             if selected_changed then
```

Comment:
Would the `reload_driver_selection` need to be called for this `selected_changed`? Or need to find out the version in the `initrd` for thorough consistency?

Reply:
My impression was that rescan is primarily for discovering installer-created file hierarchy, with depmod and the like already called. We do not really (want to) support changes to multi-version drivers behind xapi's back, so we assume if files were moved around without proper changes to the DB (not through `select`, `deselect` calls), depmod and the like would also be called. Adding logic to parse initrd too for an officially unsupported use-case might be too much.